src/target/riscv/riscv-013.c

```diff
@@ -1688,6 +1720,23 @@ static int examine(struct target *target)
 					info->index);
 			return ERROR_FAIL;
 		}
+		target->state = TARGET_HALTED;
+		target->debug_reason = DBG_REASON_DBGRQ;
+	}
+
+	/* FIXME: This is needed since register_read_direct relies on target->state
+	 * to work correctly, so, if target->state does not represent current state
+	 * of target, e.g. if a target is halted, but target->state is
+	 * TARGET_UNKNOWN, it can fail early, (e.g. accessing registers via program
+	 * buffer can not be done atomicaly on a running hart nither msattus can not
+	 * be prepared for a register access and then restored)
+	 */
+	const enum target_state saved_tgt_state = target->state;
+	const enum target_debug_reason saved_dbg_reason = target->debug_reason;
```

Comment:
I'm not sure that the debug reason would be set correctly in this case, since logic for setting it in `riscv_openocd_poll` is quite complex. We (at Syntacore) are looking to refactor `examine/poll/halt/resume` and address this in future commits.

Reply:
Almost all that complexity is to make sure the world looks as gdb needs it to. I don't think gdb cares about the debug reason when first connecting.

I just don't like this save/restore, which only exists to preserve existing behavior, which itself is already hack. It's hacks upon hacks.

If you want to set the halt reason correctly, you can probably just use:
```
				enum riscv_halt_reason halt_reason = riscv_halt_reason(target);
				if (set_debug_reason(target, halt_reason) != ERROR_OK)
					return ERROR_FAIL;
```

Reply:
I also recommend to avoid the save/restore. 

Instead, please set the target state directly based on whether we resume (or keep the hart halted) at the end of the examine.

I like Tim's suggestion: for harts that remain halted, get the debug reason via the code snippet above. It is possible that GDB does not actually need the reason upon first connect, but for consistency (and as a defensive programming measure), I recommend to do collect the reason.

(My earlier proposal with TARGET_UNKNOWN would probably work, but the suggestion above is better.)

Reply:
Addressed.

Reply:
I've checked the current code visually and it also looks good to me. Thank you.

Reply:
The change was reverted. Please see [this](https://github.com/riscv/riscv-openocd/pull/842#discussion_r1197639330)